File: sdk/java/src/main/java/com/anyvali/schemas/NeverSchema.java
package com.anyvali.schemas;

import com.anyvali.IssueCodes;
import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Schema that rejects all values.
 */
public class NeverSchema extends Schema<Object> {

    public NeverSchema() {}

    private NeverSchema(NeverSchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        String received = input != null ? input.getClass().getSimpleName() : "null";
        ctx.addIssue(IssueCodes.INVALID_TYPE,
                "No value is allowed", "never", received);
        return null;
    }

    @Override
    protected Map<String, Object> toNode() {
        return addCommonNodeFields(new LinkedHashMap<>(Map.of("kind", "never")));
    }

    @Override
    protected Schema<Object> copy() {
        return new NeverSchema(this);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/NeverSchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

class NeverSchema : Schema<Nothing>() {
    override val kind: String = "never"

    fun describe(description: String, opts: DescribeOptions? = null): NeverSchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): NeverSchema {
        applyMetadata(meta, replace)
        return this
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        return listOf(
            ValidationIssue(
                code = IssueCodes.INVALID_TYPE,
                path = ctx.path,
                expected = "never",
                received = getJsonTypeName(value)
            )
        )
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("never"))
        addMetadataToNode(this)
    }
}
